ns/Cns_getsegattrs.man
.\" Copyright (C) 1999-2002 by CERN/IT/PDP/DM
.\" All rights reserved
.\"
.TH CNS_GETSEGATTRS 3 "$Date: 2008/11/03 10:26:08 $" CASTOR "Cns Library Functions"
.SH NAME
Cns_getsegattrs \- get file segment attributes
.SH SYNOPSIS
.B #include <sys/types.h>
.br
\fB#include "Cns_api.h"\fR
.sp
.BI "int Cns_getsegattrs (const char *" path ,
.BI "struct Cns_fileid *" file_uniqueid ,
.BI "int *" nbseg ,
.BI "struct Cns_segattrs **" segattrs )
.SH DESCRIPTION
.B Cns_getsegattrs
gets the file segment(s) attributes.
This function should be called by the stager to get the tape visual identifier
and the file sequence number or blockid before executing tpread.
The file can be identified by
.I path
name or by
.IR file_uniqueid .
If both are specified,
.I file_uniqueid
is used.
.TP
.I path
specifies the logical pathname relative to the current CASTOR directory or
the full CASTOR pathname.
.TP
.I segattrs
is a pointer to an array of Cns_segattrs structures allocated by the API.
The client application is responsible for freeing the array when not needed
anymore.
.PP
.nf
.ft CW
struct Cns_segattrs {
        int             copyno;
        int             fsec;           /* file section number */
        u_signed64      segsize;        /* file section size */
        int             compression;    /* compression factor */
        char            s_status;       /* 'D' --> deleted */
        char            vid[CA_MAXVIDLEN+1];
        int             side;
        int             fseq;           /* file sequence number */
        unsigned char   blockid[4];     /* for positionning with locate */
        char            checksum_name[CA_MAXCKSUMNAMELEN+1];
        unsigned long   checksum;
};
.ft
.fi
.TP
.I copyno
is the copy number.
.TP
.I fsec
is the file segment number, It is 1 for single volume files and >= 1 for files
spanning over several tapes.
.SH RETURN VALUE
This routine returns 0 if the operation was successful or -1 if the operation
failed. In the latter case,
.B serrno
is set appropriately.
.SH ERRORS
.TP 1.3i
.B ENOENT
The referenced file does not exist anymore.
.TP
.B ENOMEM
Memory could not be allocated for the output buffer.
.TP
.B EACCES
Search permission is denied on a component of the
.I path
prefix or the caller effective user ID does not match the owner ID of the file
or write permission on the file itself is denied.
.TP
.B EFAULT
.I path
and
.IR file_uniqueid ,
.I nbseg
or
.I segattrs
is a NULL pointer.
.TP
.B ENOTDIR
A component of
.I path
prefix is not a directory.
.TP
.B EISDIR
The file is not a regular file.
.TP
.B ENAMETOOLONG
The length of
.I path
exceeds
.B CA_MAXPATHLEN
or the length of a
.I path
component exceeds
.BR CA_MAXNAMELEN .
.TP
.B SENOSHOST
Host unknown.
.TP
.B SENOSSERV
Service unknown.
.TP
.B SECOMERR
Communication error.
.TP
.B ENSNACT
Name server is not running or is being shutdown.
.SH SEE ALSO
.BR Castor_limits(4) ,
.BR Cns_chdir(3) ,
.BR Cns_setsegattrs(3)
.SH AUTHOR
\fBCASTOR\fP Team <castor.support@cern.ch>


ns/Cns_getsegattrs.c
/*
 * Copyright (C) 1999-2004 by CERN/IT/PDP/DM
 * All rights reserved
 */

/* Cns_getsegattrs - get file segments attributes */

#include <errno.h>
#include <stdlib.h>
#include <sys/types.h>
#include <unistd.h>
#include <netinet/in.h>
#include "marshall.h"
#include "Cns_api.h"
#include "Cns.h"
#include "serrno.h"
#include <string.h>

int
Cns_getsegattrs(const char *path, struct Cns_fileid *file_uniqueid, int *nbseg, struct Cns_segattrs **segattrs)
{
  char *actual_path;
  int c;
  char func[16];
  gid_t gid;
  int i;
  int msglen;
  char *q;
  char *rbp;
  char repbuf[REPBUFSZ];
  char *sbp;
  struct Cns_segattrs *segments = NULL;
  char sendbuf[REQBUFSZ];
  char server[CA_MAXHOSTNAMELEN+1];
  struct Cns_api_thread_info *thip;
  uid_t uid;
  u_signed64 zero = 0;

  strcpy (func, "Cns_getsegattrs");
  if (Cns_apiinit (&thip))
    return (-1);
  Cns_getid(&uid, &gid);

  if ((! path && ! file_uniqueid) || ! nbseg || ! segattrs) {
    serrno = EFAULT;
    return (-1);
  }

  if (path && strlen (path) > CA_MAXPATHLEN) {
    serrno = ENAMETOOLONG;
    return (-1);
  }

  if (file_uniqueid && *file_uniqueid->server)
    strcpy (server, file_uniqueid->server);
  else
    if (Cns_selectsrvr (path, thip->server, server, &actual_path))
      return (-1);

  /* Build request header */

  sbp = sendbuf;
  marshall_LONG (sbp, CNS_MAGIC4);
  marshall_LONG (sbp, CNS_GETSEGAT);
  q = sbp;        /* save pointer. The next field will be updated */
  msglen = 3 * LONGSIZE;
  marshall_LONG (sbp, msglen);

  /* Build request body */

  marshall_LONG (sbp, uid);
  marshall_LONG (sbp, gid);
  marshall_HYPER (sbp, thip->cwd);
  if (file_uniqueid && *file_uniqueid->server) {
    marshall_HYPER (sbp, file_uniqueid->fileid);
    marshall_STRING (sbp, "");
  } else {
    marshall_HYPER (sbp, zero);
    marshall_STRING (sbp, actual_path);
  }

  msglen = sbp - sendbuf;
  marshall_LONG (q, msglen); /* update length field */

  c = send2nsd (NULL, server, sendbuf, msglen, repbuf, sizeof(repbuf));
  if (c == 0) {
    rbp = repbuf;
    unmarshall_WORD (rbp, *nbseg);
    if (*nbseg == 0) {
      *segattrs = NULL;
      return (0);
    }
    segments = malloc (*nbseg * sizeof(struct Cns_segattrs));
    if (segments == NULL) {
      serrno = ENOMEM;
      return (-1);
    }
    for (i = 0; i < *nbseg; i++) {
      unmarshall_WORD (rbp, segments[i].copyno);
      unmarshall_WORD (rbp, segments[i].fsec);
      unmarshall_HYPER (rbp, segments[i].segsize);
      unmarshall_LONG (rbp, segments[i].compression);
      unmarshall_BYTE (rbp, segments[i].s_status);
      unmarshall_STRING (rbp, segments[i].vid);
      unmarshall_WORD (rbp, segments[i].side);
      unmarshall_LONG (rbp, segments[i].fseq);
      unmarshall_OPAQUE (rbp, segments[i].blockid, 4);
      unmarshall_STRINGN (rbp, segments[i].checksum_name,
                          CA_MAXCKSUMNAMELEN);
      unmarshall_LONG (rbp, segments[i].checksum);
    }
    *segattrs = segments;
  }
  if (c && serrno == SENAMETOOLONG) serrno = ENAMETOOLONG;
  return (c);
}
